### PR TITLE
[ENH] Replace `--name` and `--portal` with required arg for data description file

### LINKED ISSUE #618: support reading dataset description file for dataset level metadata
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Summary

We now read all dataset level info (current and future) from a dataset description file that is compatible with / aligned with the BIDS dataset_description.json and specced here: https://github.com/neurobagel/planning/issues/335

Example:
```json
{
  "Name": "My Dataset",
  "Authors": ["First Author", "Second Author"]
  "ReferencesAndLinks": [
    "https://mydataset.org", 
    "https://mydatasetportal.org"
  ],
  "Keywords": ["fMRI", "PD"],
  "RepositoryURL": "https://datasets.datalad.org/mydataset",
  "AccessInstructions": "Submit a data access request at https://mydataset.org.",
  "AccessType": "restricted",
  "AccessEmail": "first.author@gmail.com",
  "AccessLink": "https://mydatasetportal.org",
}
```

TODO:

- [x] Add new file input argument for `bagel pheno` to receive required file `dataset_description.json`
- [x] Validate loaded `dataset_description.json` against schema
- [x] Ensure empty or whitespace only strings for fields are ignored
- [x] Add informative error if `dataset_description.json` is schema invalid (e.g., missing required key)
  - [ ] and point to https://github.com/neurobagel/documentation/issues/348 docs
- [x] If file is valid but is missing some  / all of our optional keys, emit a warning to tell them how they could add these keys
- [x] Update the pydantic data model for the pheno.jsonld output
  - [x] Deprecate `hasPortalURI`
      - **NOTE: This will mean that users can no longer add BIDS or derivative metadata to phenotypic-only JSONLDs generated with older CLI versions - they must update the phenotypic JSONLD file first (even though these old files will still work with the n-API)**
- [x] Make and test that CLI can create updated pheno.jsonld
- [x] Remove the `--name` and `--portal` arguments and all related validators, tests, and test fixtures

### Documentation update(s) needed

- [x] Add helptext context for new `dataset_description.json` input argument that also explains how to make file / reuse BIDS file.

### PR DESCRIPTION
**Pull request by alyssadai**
- Closes #618

> [!NOTE]
> - https://github.com/neurobagel/neurobagel_examples/pull/89 should be merged first - this PR currently checks out the relevant PR branch of the submodule to confirm tests pass

Changes proposed in this pull request:

- Replace required `--name` and optional `--portal` args for `bagel pheno` with required `--dataset-description` arg
- Emit informative error if dataset description is schema invalid
- Emit warning about any unset/empty optional fields for a valid dataset description to encourage adding all info
- Add logic to handle schema-valid dataset description fields that are empty (e.g., `""` or `[]`) or have all-whitespace values (e.g., `"  "`) so they are treated the same as unset fields (without erroring), ensuring:
  - Warnings about missing recommended fields are complete
  - Dataset names (required) are not empty strings
  - JSONLD doesn't end up with unmeaningful dataset metadata that will be hard to detect/filter out later
- Update JSONLD data model and `bagel pheno` command to add new dataset metadata
- Refactor out saving of final JSONLD to ensure all commands use same method
- Bump min Pydantic version to be able to disable normalization of raw URLs from dataset description files (see https://github.com/pydantic/pydantic/issues/7186)

## Checklist
_This section is for the PR reviewer_

- [x] PR has an interpretable title with a prefix (`[ENH]`, `[FIX]`, `[REF]`, `[TST]`, `[CI]`, `[MNT]`, `[INF]`, `[MODEL]`, `[DOC]`) _(see our [Contributing Guidelines](https://neurobagel.org/contributing/CONTRIBUTING#pull-request-guidelines) for more info)_
- [x] PR has a label for the release changelog or `skip-release` (to be applied by maintainers only)
- [x] PR links to GitHub issue with mention `Closes #XXXX`
- [x] Tests pass
- [x] Checks pass

For new features:
- [x] Tests have been added

For bug fixes:
- [ ] There is at least one test that would fail under the original bug conditions.

## Summary by Sourcery

Validate and log completeness of dataset description metadata when running the pheno CLI, and extend internal models to capture dataset-level metadata and access information.

New Features:
- Add a required --dataset-description/--m option to the pheno CLI for supplying dataset metadata JSON.
- Introduce validation utilities for dataset_description, including detection of missing or whitespace-only optional fields and logging of informative warnings instead of failing.

Enhancements:
- Extend the Dataset Pydantic model to include authors, homepage, keywords, repository URL, access instructions, access type, access email, and portal URI fields derived from dataset description metadata.
- Tighten dataset description URL types to HttpUrl and enforce non-whitespace Name plus normalization of empty/whitespace metadata values via Pydantic validators.
- Improve validate_inputs docstring for clarity.

Tests:
- Add unit tests covering valid, incomplete, whitespace-only, and invalid dataset_description payloads and their corresponding warnings or errors.

## Summary by Sourcery

Require a dataset description JSON file for the pheno CLI, validate and surface dataset-level metadata and access information in the JSON-LD output, and centralize JSON-LD serialization across commands.

New Features:
- Add a required --dataset-description/--m option to the pheno CLI that loads dataset metadata from a JSON file instead of separate name and portal arguments.
- Introduce a DatasetDescription Pydantic model and utilities to validate dataset description payloads and derive dataset-level attributes for JSON-LD export.

Enhancements:
- Extend the Dataset model and JSON-LD schema to include authors, references/links, keywords, repository URL, and access-related fields derived from the dataset description.
- Normalize and validate dataset description fields (including URL, email, and whitespace handling) and emit warnings when recommended metadata fields are unset or empty.
- Refactor JSON-LD creation into a reusable dataset_to_jsonld helper and apply it across pheno, bids, and derivatives commands.
- Ensure JSON-LD files are written with UTF-8 encoding and non-ASCII characters preserved by default.
- Relax handling of legacy hasPortalURI on Dataset to maintain backward compatibility while preferring richer access metadata.

Build:
- Bump the minimum Pydantic version to >=2.12 to support updated URL handling and validators.

Tests:
- Update CLI and logging tests to use the new --dataset-description argument in place of name/portal flags.
- Add unit tests covering dataset description validation, including incomplete, whitespace-only, valid, and invalid payloads, plus context/JSON-LD generation with the new helper.

### CODE REVIEW

**Comment by alyssadai** [comment]
@sourcery-ai review

File: bagel/utilities/file_utils.py [hunk 1]
```diff
@@ -95,8 +95,8 @@ def check_overwrite(output: Path, overwrite: bool):
 
 
 def save_jsonld(data: dict, filename: Path):
-    with open(filename, "w") as f:
-        f.write(json.dumps(data, indent=2))
+    with open(filename, "w", encoding="utf-8") as f:
+        f.write(json.dumps(data, indent=2, ensure_ascii=False))
```
Comment (surchs) [inline bagel/utilities/file_utils.py]: is this in response to a bug? If so, we should probably add a test for whatever the bug was about

File: bagel/utilities/model_utils.py [hunk 2]
```diff
@@ -32,13 +32,12 @@ def generate_context(config: str) -> dict:
...
-def add_context_to_graph_dataset(dataset: models.Dataset, config: str) -> dict:
-    """Generate and add the Neurobagel context to a graph-ready dataset to form a JSONLD dictionary."""
-    context = generate_context(config)
+def dataset_to_jsonld(context: dict, dataset: models.Dataset) -> dict:
+    """Add the Neurobagel context to a graph-ready dataset to form a JSONLD dictionary."""
     # We can't just exclude_unset here because the identifier and schemaKey
     # for each instance are created as default values and so technically are never set
     # TODO: we should revisit this because there may be reasons to have None be meaningful in the future
-    return {**context, **dataset.model_dump(exclude_none=True)}
+    return {**context, **dataset.model_dump(exclude_none=True, mode="json")}
```
Comment (surchs) [inline bagel/utilities/model_utils.py]: Out of curiosity: why is this now only about merging the context with the model dump?

File: bagel/dataset_description_model.py [hunk 3]
```diff
@@ -82,12 +91,46 @@ class DatasetDescription(BaseModel):
...
+    @field_validator("access_instructions")
+    @classmethod
+    def whitespace_string_to_default_none(
+        cls, value: str | None
+    ) -> str | None:
+        """Convert an empty string or a string that contains only whitespace to None."""
+        if value is not None and value.strip() == "":
+            return None
+        return value
+
```
Comment (surchs) [inline bagel/dataset_description_model.py]: so iiuc, this is one way for you to enable a fallback default value if the user provides a nonsensical input (e.g. empty string)? I'm wondering: what will we do with this in practice? If the user added "access_instructions" as empty string, and we don't warn them about it / error out, what are we going to do with the info?

**Review by surchs — APPROVED** [review]
Thanks @alyssadai for taking this on. This all looks very nice. I have two questions: 1) about something that looks like a bugfix without a bug, and 2) about a possibly missing test for URL parsing (may have missed it)

Take a look, and then gtg 🧑‍🍳

File: bagel/models.py [hunk 4]
```diff
@@ -21,7 +22,10 @@ class Bagel(BaseModel):
         ),
     ]
 
-    model_config = ConfigDict(extra="forbid")
+    # NOTE: url_preserve_empty_path (>=2.12) is needed to prevent HttpUrl from auto-appending trailing slashes to URLs
```
[the remark this answers is not in the document]
Comment (alyssadai) [reply bagel/models.py]: Yes, it's referring to the Pydantic version where this config parameter was introduced.

File: bagel/utilities/file_utils.py (hunk 1 above)
Comment (alyssadai) [reply bagel/utilities/file_utils.py]: Not a bug, I just noticed when testing the output that currently non-ascii characters like é are escaped (`\u00e9`) in the final jsonld. Afaik this shouldn't have any functional consequence on downstream tools since the characters are decoded into Unicode again as soon as they're read in by Python (e.g., in n-API responses) but they're just less human-readable in the raw file.

File: bagel/dataset_description_model.py (hunk 3 above)
Comment (surchs) [reply bagel/dataset_description_model.py]: Thanks @alyssadai. I think that is definitely enough then to let the user know they should probably fix the file!
